src/Components/Components/src/Rendering/RendererSynchronizationContext.cs

```diff
@@ -63,7 +63,7 @@ public Task Invoke(Action action)
         public Task InvokeAsync(Func<Task> asyncAction)
         {
             var completion = new TaskCompletionSource<object>();
-            Post(async (_) =>
+            ExecuteSynchronouslyIfPossible(async (_) =>
```

Comment:
Unrelated to this change, but how important is perf on these code paths?  Right now this is allocating a delegate and a closure on every call, in order to capture both asyncAction and completion.  You could avoid both of those by creating your own TCS-derived type, e.g.
```C#
internal sealed class StatefulTaskCompletionSource<TState, TResult> : TaskCompletionSource<TResult>
{
    public TState State;
}
```
and then passing that object into the callback via its `object state`, e.g.
```C#
var completion = new StatefulTaskCompletionSource<Func<Task>, object>() { State = asyncAction };
ExecuteSynchronouslyIfPossible(async state =>
{
    var local = (StatefulTaskCompletionSource<Func<Task>, object>)state;
    try
    {
        await local.State();
        local.SetResult(null);
    }
    catch (OperationCanceledException)
    {
        local.SetCanceled();
    }
    catch (Exception exception)
    {
        local.SetException(exception);
    }
}, completion);
return completion.Task;
```

Reply:
We'd love to improve the perf of this. I'll give your idea a try.